### ISSUE TITLE
[Bug] TypeScript issue when using GeoJsonLayerProps type for FeatureCollection

### ISSUE THREAD

**Issue by krishamehta903** [visgl/deck.gl#7571]
### Description

We are using GeoJsonLayer's in our deck.gl application with TypeScript, where we have confirmed that out data input for these layers follow the GeoJSON specifications (and are typed with the `GeoJSON` types from the geojson package). The data we are passing in is a feature collection of several polygons (features) - however when we try to use `GeoJsonLayerProps` from @deck.gl/layers/typed, we get TS errors that the GeoJSON data does not confirm to the type spec for `data`:

`Type 'GeoJSON' is not assignable to type 'string | LayerData<Feature<Geometry, { [name: string]: any; }>> | AsyncIterable<Feature<Geometry, { [name: string]: any; }>> | Promise<...>'.`

It appears that the GeoJsonLayerProps accounts for single feature types but not feature collection types.

### Flavors

- [X] React
- [ ] Python/Jupyter notebook
- [ ] MapboxLayer
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] DeckLayer/DeckRenderer for ArcGIS

### Expected Behavior

We expect to not see these TypeScript errors when passing in a collection of features into a geojson layer and using the GeoJsonLayerProps from deck.gl

### Steps to Reproduce

```
import { GeoJsonLayer, GeoJsonLayerProps } from '@deck.gl/layers/typed'
import type { GeoJSON } from 'geojson'

export const buildGeoJsonLayer = (
  props: Partial<GeoJsonLayerProps>
): GeoJsonLayer => {
  return new GeoJsonLayer({
    autoHighlight: true,
    data: '',
    extruded: false,
    filled: true,
    getLineColor: [0, 0, 0],
    getStrokeColor: [255, 255, 255],
    highlightColor: [0, 138, 216],
    lineWidthMinPixels: 1,
    lineWidthMaxPixels: 1,
    opacity: 0.6,
    pickable: true,
    stroked: true,
    visible: true,
    wireframe: true,
    ...props,
  })
}

const myData: GeoJSON = {type: "FeatureCollection", "features": [{"geometry": {"type: "Polygon", coordinates: ...}, "type": "Feature"}, ...<insert additional feature objects here>...]}

  const layers = [
    buildGeoJsonLayer({
      id: 'my-geojson-layer',
      data: myData,
      onClick: onFixtureClick,
      getElevation: (f: Feature) => f.properties?.height || 3,
      getFillColor: (f: Feature) => (f === selectedFeature ? BLUE : GRAY),
      updateTriggers: {
        getFillColor: [selectedFeature],
      },
    })
]
```

Notice that there's a TS error when assigning `myData` to `data` using the `buildGeoJsonLayer` util function to populate the `layers` array.

### Environment

- Framework version: deck.gl@8.8.21
- Browser: Chrome 108.0.5359.124
- OS: macOS 12.6.1

**Comment by Pessimistress** [comment]
@zbigg

Right now the layer generic is defined as `Layer<DatumT>` where `props.data` is expected to be an iterable of `DatumT` (among other things). It is not possible to override the type of `props.data`.

We could try to define it as `Layer<DataT>` where `props.data` is `DataT` instead? In accessors we could still access the element type as

```typescript
type DatumT = DataT extends Iterable<any> ? DataT[0] : null;
```

**Comment by zbigg** [comment]
> We could try to define it as Layer<DataT> where props.data is DataT instead?
> type DatumT = DataT extends Iterable<any> ? DataT[0] : null;

@Pessimistress 
Do you mean, that `DataT` would be `Feature[] | FeatureCollection` for geojson layer?

If yes, i am not sure if i understand how it is supposed to wkrk, but i worry that wouldn't help either. `FeatureCollection` still has "other", very custom "feature" type - it would be `DataT['features']` and i have no idea how to pass this kind of "type accessor" down to base layer.

Looking from other direction, i don't have any easy fix for this.
`geojson-layer` basically breaks assumption that there that "rows" passed in `data` can be "carried" by collection different than array/iterable.
The only type-correct, solution i can think of now would be do add `DataContainerT = Iterable<DataT> ...` generic param to every layer-data related type ... and customize it only for `geojson-layer.` Huge change at first look.

**Comment by zbigg** [comment]
There is a way to "override" type of `data` in simple sandbox-like experiments like [this](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYhCGwCuAnaBeKBvKpIC4oByORVCIgGigHMIB7AWwmBREIDslGAjCFANoBdamBT1IKYAEsIAZ0JYAvkoBQeaKWRoAwvQA2+iAGMZ9DlEw4NhEgm0Q9hk2Y5UoAM3vkFsb2mE1dXBoABl4EH4AEUR4AB4AFQA+SygEqAAfNOEoVWDIKHDIlAAFcTA5RJSrVSg6qGkAE0I5VmkOGlr6xtiAfkIi6Niq1SCNKAAhVIB5RmlgOMHS8sqtciTqIh7geCIUgDJc+uwobfgBiKGduLW0FKzbxwMjU2lzKCDjc1bPQkfU6whWyPdx0JgsNiEYSicr8GTyRRQc7EXYqVRfDg-DzGP7+J7OV7vKy4IHER5OF6udxeMhoXyBPIYn4AD0IUxqxyatmZIAAXu4APQCqCtFDtGiecSMLp1M6EbFQIVQYzwCx8JFyKAQZmQUwQRp+WnQB54nImo0UlxvDijPLMgB0ZyAA), but i'm not sure if it will be possible and/or feasible to apply to current code base.

**Comment by Pessimistress** [comment]
Closed by #7682
